Feature: Custom Application Rules System

kindingerlek opened the pull request:
# Custom Application Rules System
This PR introduces a custom application rules system that lets users control how individual applications interact with Tiling Shell's features. 

 - fix #200,
 - fix #226 
 - fix #334;

<img width="3440" height="1440" alt="image" src="https://github.com/user-attachments/assets/1f8fd27e-4a96-4785-9856-7da1adcc03f0" />


## What This Adds
**Per-Application Configuration**:
Users can now create rules for specific applications (identified by App Id) to enable or disable:

 - Custom window borders
 - Auto-tiling
 - Snap assistant
 - Window suggestions
 - Resize complementing windows
 - Span multiple tiles
 - Automatic Fullscreen Handling : A built-in default rule automatically disables all tiling features when any application goes fullscreen, ensuring a clean fullscreen experience.

**Preferences UI** 
A new settings panel allows users to:

 - Add custom rules for any application;
 - Edit existing rules through a clean modal interface;
 - Delete rules when no longer needed;
 - Add an application through a list to easily check what’s installed on the system, with a manual input option if the application picker can’t find it.

<img width="610" height="677" alt="image" src="https://github.com/user-attachments/assets/2aa0ffb1-a9cc-4173-97cf-5bc86c817b58" />


## Implementation

Created three new components:

 - `ApplicationPicker`: Helps users to locate the installed applications in your system. Returns the `applicationId`;
 - `CustomRulesManager`: Evaluates rules and determines which features should be active for each window;
 - `CustomApplicationRulePrefs`: Provides the settings UI for managing rules;

Integrated the rule system into:

 - `TilingManager` (auto-tiling, snap assist, window suggestions, span multiple tiles)
 - `ResizingManager` (resize complementing)
 - `WindowBorderManager` (custom borders)

**User Experience**
Rules are organized in a modal dialog with grouped settings for clarity. The system validates input to prevent duplicate entries and provides clear visual feedback. Default behavior remains unchanged when no custom rules are defined.

## No Breaking Changes
This is purely additive - existing functionality works exactly as before for users who don't configure custom rules.

## Future Ideas
The custom rules foundation opens possibilities for more granular per-application control:

 - **Custom gaps**: Override inner/outer gap sizes for specific applications
 - **Border styling**: Custom border colors, widths, or radius per application
 - **Layout preferences**: Assign specific layouts or tile positions to applications
 - **Window size constraints**: Set minimum/maximum tile spans for certain apps
 - **Startup behavior**: Control initial window placement when applications launch
 - **Workspace rules**: Automatically move applications to specific workspaces
 - **Keyboard shortcut overrides**: Per-app keybinding customization

kindingerlek committed:
feat: adds a settings panel and scheme for blacklist applications

kindingerlek committed:
feat: expand settings for blacklisted apps

kindingerlek committed:
refact: rename from blacklist to custom rules

kindingerlek committed:
feat: integrate CustomRulesManager into TilingManager and extension initialization

kindingerlek committed:
feat: integrate CustomRulesManager into ResizingManager and TilingManager for enhanced window management

kindingerlek committed:
feat: enhance CustomRulesManager with fullscreen default rule and refactor application config

kindingerlek committed:
refact: rename CustomRulesApplicationConfig to ConfigRules and update references

kindingerlek committed:
refact: remove broken accordion behavior implementation from TilingShellExtensionPreferences

kindingerlek committed:
refact: create a separated file for managing the custom application rule preferences

kindingerlek committed:
refact: migrate from a collapse row to a modal, with better grouping for rules

kindingerlek committed:
feat: refactor custom application rules to use application IDs instead of WM classes and add application picker dialog

domferr commented on src/components/applicationPicker.ts:
```diff
@@ -0,0 +1,341 @@
+
+        // Remove organization prefixes (e.g., org.gnome.Nautilus -> nautilus)
+        const parts = appId.split('.');
+        return parts[parts.length - 1];
```
This code is not reachable due to the previous return (maybe that is not needed?)

kindingerlek replied:
You are right. This is a left over from previous attempts.

kindingerlek replied:
You are right. This is a left over from previous attempts. I'll remove it. Thanks

domferr commented on src/components/customRulesManager.ts:
```diff
@@ -0,0 +1,223 @@
+        Signals: {
+            'customRules-changed': {},
+        },
+    };
```
With the latest version I have modified the GObject registration. This code becomes
```
export class CustomRulesManager extends GObject.Object {
    static { registerGObjectClass(this, {
        GTypeName: 'CustomRulesManager',
        Signals: {
            'customRules-changed': {},
        },
    });
    ...
}
```

domferr committed:
set version to 17.3

albanobattistella committed:
Update Italian translations in it.po (#466)

Updated Italian translations for various strings.

ChrisLauinger77 committed:
Updates German translation (de) (#448)

Updates the German translation file with new and updated
translations. This includes translations for:

- Adding tiled windows to ALT+TAB menu
- Edge tiling offset
- Restricting directional focus to tiled windows
- Choosing custom color
- Use system accent color
- Cycle layouts

This ensures that the user interface is properly translated
into German for a better user experience.

alewicki95 committed:
Update pl.po and tilingshell@ferrarodomenico.com.pot from source (#445)

* Update tilingshell@ferrarodomenico.com.pot

* Update pl.po

EkaterinePapava committed:
po: Add Georgian translation (#457)

taijuin committed:
Add files via upload (#453)

domferr committed:
add gresources support to preferences and create row for edge tiling behaviour

domferr committed:
feat: add Wacom tablet support

domferr committed:
feat: reorder layouts

kindingerlek committed:
refact: remove deadcoded

kindingerlek committed:
feat: enhance import handling to support @ path aliases and improve error logging in legacy file processing

kindingerlek commented:
Hello @domferr, I rebased the branch onto `v17.3` and resolved the conflicts. I believe it’s now ready to be merged. Thank you for accepting it, and thanks again for this awesome extension.

kindingerlek committed:
feat: enhance import handling to support @ path aliases and improve relative path resolution

kindingerlek committed:
feat: update CustomRulesManager to register GObject class correctly and add emit method for TypeScript compatibility

lousuan committed:
feat: Update Simplified Chinese Translation (#468)

jessedhillon committed:
feat: add sync layout option to snap assist (#469)

Gregorein committed:
Adaptive corner/edge snap (#323)

* feat: implement edge snap mode settings and behavior

* refactor: update edge snap mode options in preferences

- Changed edge snap mode descriptions to more accurately reflect functionality

* rebasing supporting latest changes

* feat: implement edge snap mode settings and behavior

* fix imports

* feat: add modes icon instead of dropdown menu

* update translations

---------

Co-authored-by: Domenico Ferraro <ferraro.domenico125@gmail.com>

jessedhillon committed:
chore: on GNOME 49+, use Meta.Monitor.get_display_name() (#470)

* chore: on GNOME 49, use Meta.Monitor.get_display_name()

* fix: logical monitor doesn't have get_layout

domferr committed:
fix: ensure the border is on top of the window

domferr committed:
some readability improv

domferr committed:
Merge branch 'v17.3' into feat/custom-rules

domferr committed:
feat: raise all windows together

domferr committed:
Merge branch 'v17.3' into feat/custom-rules

domferr committed:
fix: use quarters windows suggestions with default edge tiling

domferr committed:
multiple refactoring for edge snapping

domferr committed:
refactor: use Clutter.ModifierType as mask

domferr committed:
fix: window borders scaling factor

domferr committed:
Merge branch 'v17.3' into feat/custom-rules

SimonPrague commented:
I can't wait for this!
